pkg/backend/httpstate/journal/snapshot_test.go

```diff
@@ -341,3 +343,217 @@ func TestJournaler413MinimumBatchSize(t *testing.T) {
 	err := journaler.Close()
 	require.NoError(t, err)
 }
+
+// mockTicker is a manually controlled ticker for testing.
+type mockTicker struct {
+	c      chan time.Time
+	mu     sync.Mutex
+	period time.Duration
+}
+
+func newMockTicker() *mockTicker {
+	return &mockTicker{
+		c: make(chan time.Time, 1),
+	}
+}
+
+func (m *mockTicker) C() <-chan time.Time {
+	return m.c
+}
+
+func (m *mockTicker) Stop() {
+	// no-op for testing
+}
+
+func (m *mockTicker) Reset(d time.Duration) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.period = d
+}
+
+func (m *mockTicker) Tick() {
+	m.c <- time.Now()
+}
+
+// TestSendBatchesOneBatchAtATime ensures only one batch is being sent at a time.
+func TestSendBatchesOneBatchAtATime(t *testing.T) {
+	t.Parallel()
+
+	entries := make(chan saveJournalEntry, 10)
+	tick := newMockTicker()
+
+	// Track concurrent sends
+	var activeSends int32
+	var maxConcurrentSends int32
+	var sendCount int32
+	var mu sync.Mutex
+	sendDelay := 50 * time.Millisecond
+
+	sender := func(batch []apitype.JournalEntry) error {
+		current := atomic.AddInt32(&activeSends, 1)
+		defer atomic.AddInt32(&activeSends, -1)
+
+		mu.Lock()
+		if current > maxConcurrentSends {
+			maxConcurrentSends = current
+		}
+		atomic.AddInt32(&sendCount, 1)
+		mu.Unlock()
+
+		// Simulate slow send to make concurrency issues more apparent
+		time.Sleep(sendDelay)
+
+		return nil
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		sendBatches(1, 100*time.Millisecond, entries, sender, tick)
+	}()
+
+	results := make([]chan error, 3)
+	for i := 0; i < 3; i++ {
+		results[i] = make(chan error, 1)
+		entries <- saveJournalEntry{
+			entry:  apitype.JournalEntry{SequenceID: int64(i)},
+			result: results[i],
+		}
+	}
+
+	for _, result := range results {
+		<-result
+	}
+
+	close(entries)
+	<-done
+
+	assert.Equal(t, int32(1), maxConcurrentSends, "Only one batch should be sent at a time")
+	assert.Equal(t, int32(3), sendCount, "Expected 3 batches to be sent")
+}
+
+// TestSendBatchesSendsAfterTimerTick ensures batches are sent when the timer ticks.
+func TestSendBatchesSendsAfterTimerTick(t *testing.T) {
+	t.Parallel()
+
+	entries := make(chan saveJournalEntry, 10)
+	tick := newMockTicker()
+
+	var batchesSent int32
+	batches := [][]apitype.JournalEntry{}
+
+	sender := func(batch []apitype.JournalEntry) error {
+		batches = append(batches, batch)
+		atomic.AddInt32(&batchesSent, 1)
```

Comment:
Why is this atomic, when the update to `batches` isn't? If we need atomic then we need to protect `append`. If not, then we can just use `batchesSent++`.